### ISSUE TITLE
Add support for Kubernetes NodePort

### ISSUE THREAD

**Issue by michi-covalent** [cilium/cilium#869]
My understanding based on a conversation with @tgraf is that cilium currently doesn't support NodePort.

**Comment by tgraf** [comment]
Documentation:
https://kubernetes.io/docs/concepts/services-networking/service/#type-nodeport

Details:
- A NodePort service is a service which is reachable on each worker node on a particular port (allocated by kubernetes or specified by the user)
- Cilium can easily provide this configuring an LB on each node with L3=NodeIP, L4=nodePort
- bpf_netdev will perform LB on any external traffic coming in to a nodePort
- bpf_lxc will perform LB of any local pods talking to a nodePort traffic
- It looks like no changes on the BPF are required

TODO:
- [ ] Convert type `NodePort` services into a regular cilium service (LB) with L3=NodeIP, L4=
```
type L3n4Addr struct {
        IP: <node IP>
        L4Addr: [*].nodePort
}
```
- [ ] Add tests

**Comment by tgraf** [comment]
This work has been completed.

**Comment by anitgandhi** [comment]
Sorry to bring up an old issue, but is NodePort officially implemented? 

I see this PR (https://github.com/cilium/cilium/pull/1171) was closed, presumably in favor of this one (https://github.com/cilium/cilium/pull/1259), which remains open. So I wasn't sure what the status was.

**Comment by tgraf** [comment]
> Sorry to bring up an old issue, but is NodePort officially implemented?
> 
> I see this PR (#1171) was closed, presumably in favor of this one (#1259), which remains open. So I wasn't sure what the status was.

@anitgandhi Yes. We have implemented full compatibility with either IPVS or iptables based kube-proxy to provide NodePort. We have not implemented a BPF based NodePort  implementation yet.

**Comment by anitgandhi** [comment]
Gotcha! Yeah the latest k8s 1.11 kube-proxy in IPVS mode worked fine for me with cilium 1.1; just wanted to make sure I wasn't missing something. 

Thank you for the quick response!

P.S. if/when the BPF based nodeport happens, will that effectively remove the need for kube-proxy?

**Comment by tgraf** [comment]
> P.S. if/when the BPF based nodeport happens, will that effectively remove the need for kube-proxy?

Correct. kube-proxy is already out of the picture for all pod to pod communication.

**Comment by nathanjsweet** [comment]
@tgraf 

What's the status of this issue? It looks like it ended up not moving forward at all.

I'd be interested in picking this up and moving it forward if there aren't sufficient resources to complete it.

**Comment by tgraf** [comment]
> What's the status of this issue? It looks like it ended up not moving forward at all.

Support for this got merged into the 1.6 branch. Fixed in #8384
